sequana/ribodesigner.py

```diff
@@ -12,101 +12,253 @@
+            # Starts reverse probes to be centered on inter_probe_space of the forward probes
+            starts = [int((starts[i + 1] + starts[i]) / 2) for i in range(0, len(starts) - 1)]
+            stops = [start + probe_len for start in starts]
+
+        else:
+            sequence = seq.sequence
+
+        df = pd.DataFrame({"name": seq.name, "start": starts, "stop": stops, "strand": strand, "score": 0})
+        df["sequence"] = [sequence[row.start : row.stop] for row in df.itertuples()]
+        df["seq_id"] = df["name"] + f"_{strand}_" + df["start"].astype(str) + "_" + df["stop"].astype(str)
+
+        if strand == "-":
+            # Transform to bed coordinates for the reverse_complement
+            df["start"] = len(sequence) - df["start"]
+            df["stop"] = len(sequence) - df["stop"]
+            df.rename(columns={"start": "stop", "stop": "start"}, inplace=True)
+
+        return df
 
-    logger.info(f"{count} probes designed.")
+    def get_all_probes(self):
+        """Run all probe design and concatenate results in a single DataFrame."""
 
+        probes_dfs = []
 
-def cluster_probes(fasta_in, fasta_out, seq_id_thres=0.80, threads=4):
-    """Use cd-hit-est to cluster highly similar probes
-
-    :param fasta_in: path to FASTA file with probes.
-    :param fasta_out: path to FASTA file with probes after filtering.
-    """
-    log_file = Path(fasta_out).with_suffix(".log")
-
-    cmd = f"cd-hit-est -i {fasta_in} -o {fasta_out} -c {seq_id_thres} -n {threads}"
-    logger.info(f"Clustering probes with command: {cmd} (log in '{log_file}').")
-
-    with open(log_file, "w") as f:
-        subprocess.run(cmd, shell=True, check=True, stdout=f)
-
-
-def fasta_to_csv(fasta_in, csv_out):
-    """Convert a FASTA file into a CSV file.
-
-    The oligo names are in column 1 and the oligo sequences in column 2.
-
-    :param fasta_in: path to FASTA file with clustered probes.
-    :param csv_out: path to CSV file with clustered probes.
-
-    """
-
-    seq_dict = {"id": [], "sequence": []}
+        with pysam.FastxFile(self.ribo_sequences_fasta) as fas:
+            for seq in fas:
 
-    with pysam.FastxFile(fasta_in) as fas:
-        data = [(seq.name, seq.sequence) for seq in fas]
-    logger.info(f"After clustering, keeping {len(data)} probes.")
-    logger.info(f"Creating probes csv table '{csv_out}'.")
-    pd.DataFrame(data, columns=["id", "sequence"]).to_csv(csv_out, index=False)
+                probe_len, step_len = self._get_probe_and_step_len(seq)
+
+                probes_dfs.append(self._get_probes_df(seq, probe_len, step_len, "+"))
+                probes_dfs.append(self._get_probes_df(seq, probe_len, step_len, "-"))
+
+        probes_df = pd.concat(probes_dfs)
+        self.probes_df = probes_df
+
+    def export_to_fasta(self):
+        """From the self.probes_df, export to FASTA and CSV files."""
+
+        with open(self.probes_fasta, "w") as fas:
+            for row in self.probes_df.itertuples():
```

Comment:
Je crois qu'un zip irait plus vite en mode
```
for seq_id, seq int zip(self.probes_df['seq_id'], self.probes_df['sequence']):

Reply:
C'est noté pour les suivantes, mais là je pense que je vais laisser en l'état (au maximum on écrit ~400 sequence et une seule fois)